### ISSUE TITLE
Package did not load correctly.

### ISSUE THREAD

**Issue by Riobe** [quetzalcoatl/xvsr10#1]
http://i.imgur.com/TQcV7.png

I downloaded the VSIX 'installer' file and ran it, according to the directions in the ReadMe. Upon opening a solution I or trying to open the 'xUnit Runner manager' I get the error in the image above. If I try to debug a test, I get the following output:

E, 5184, 11, 2012/05/07, 00:11:18.131, RIOBE\QTAgent32.exe, AgentObject: Exception while loading test adapter: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.VisualStudio.TestTools.Execution.TestAdapterManager.CreateAdapter(String adapterTypeName)
   at Microsoft.VisualStudio.TestTools.Execution.TestAdapterManager.LoadAndInitAdapter(String adapterTypeName, ITestElement testElement, IRunContext runContext, Boolean shouldPrepare)
   at Microsoft.VisualStudio.TestTools.Execution.TestAdapterManager.LoadAdapter(ITestElement testElement, IRunContext runContext, Boolean checkHostAdapter, Boolean shouldPrepare)
   at Microsoft.VisualStudio.TestTools.Agent.AgentObject.ExecuteTest(ITestElement test, IResultSink resultSink, TestContext testContext, Boolean loadTest, Boolean forceSynchronousExecution)
The thread 'Agent: test queue thread' (0x12e8) has exited with code 0 (0x0).

I am on Windows 7 64 bit running Visual Studio 2010 Ultimate. The project in question is referencing xUnit 1.9 dlls and has the MSTest GUID in its ProjectType section. I can't think of any other relevant details, but will update if they are requested.

**Comment by quetzalcoatl** [comment]
Thanks for the first issue! :)

I cannot tell anything definitely, but it looks like the 'TestAdapter' class was not found. Due to the architecture of MSTest, it is loaded by typname/assemblyname and the xvsr10 dlls must be placed in one of three possible directories (well, or the GAC, but I'd rather not try that).
- Is it possible that your VS installation place is nonstandard or manually moved etc? anything other than typical c:\programfiles\microsoftivsualstudioXYZ ?
- After you have installed the VSIX and opened some solution, did a "installer" window show up? did all files install correctly?

If you have some spare time, you can check it manually by opening "test" menu and running 'xUnit runner manager". once run, it should ask about UAC permissions if everything went well, it should look like: http://imgur.com/wC3FV . Please check that the 'VisualStudio instance" is detected well and that the two rows at the bottom are similar to the screenshot. Also, please check that the QTAgent config file was selected properly. On my screenshot that's QTAgent32.exe.config, because I have x86 platform. Your QTAgent config file will be probably different.

Hm. Speaking of that. You say to have 64bit platform, but the QTAgent error says QTAgent32.exe.
Is your project forced to compile/run in x86 mode? After thinking about it, I'm almost sure you are.

If so, that's the problem. When writing the installer. I've not thought of that option. The "installer" detects your platform and patches the QTAgent config files with respect to some settings that the VS returns. If your platform is x64, then probably the QTAgent.exe.config was patched and but your project setup seems to use QTAgent32.exe.config. **Please confirm that**!

If this is the case, then I have to widen the installation a bit and patch all config files for QTAgent, not only the "thought-to-be-proper" one.
This is simple to fix, but will take me a few days, as I won't have much time at least until thursday.

If you can't wait, a very quick manual patch can be made:
- open QTAgent32.exe.config
- open QTAgent.exe.config
- copy two "dependentAssembly" tags (for xunit.runner.utility and xunit.runner.visualstudio.vs2010.core) so the "assemblyBinding" section looks similar in both files

**Comment by Riobe** [comment]
Heheh, thanks for the fast reply! I was at work so I couldn't reply till now. My Visual Studio is a standard install. I left it default. It means some things are in C:\Program Files\Microsoft Visual Studio 10.0... and some are in Program Files (x86)\Microsoft Visual Studio 10.0...

After I installed the VSIX I get an error immediately upon opening a solution with a test project or trying to open the "xUnit runner manager". I can't actually open the runner manager because of that, so I'm not sure if it would look like the screenshot you posted.

I hope I say this right: I'm trying to test a class library that offers some functionality that relies on XNA (an InputManager at the moment). XNA assemblies are x86, so I can't reference them and run in 64 bit mode. So while my machine is a 64 bit machines, I have projects set to build to x86. I found the config files for QTAgent.exe and QTAgent32.exe, but I'm not sure what your 3rd bullet point means. Are there a config files for xunit.runner.utility and xunit.runner.visualstudio.vs2010.core somewhere that I can modify?

Also, although I get that error right when I open one of the solutions I had that already had test projects, I opened one that didn't and added a test project to it. Then I hit Ctrl+R,A (Run all tests in Solution) and got this result: 

The test adapter ('Xunit.Runner.VisualStudio.VS2010.XUnitTestAdapter, xunit.runner.visualstudio.vs2010.core, Version=2.1.0.0, Culture=neutral, PublicKeyToken=8caa1cd9f64d6a1a') required to execute this test could not be loaded. Check that the test adapter is installed properly. Object reference not set to an instance of an object.

This is making me think that just running the VSIX wasn't enough and I need to put this xunit.runner.visualstudio.vs2010.core assembly somewhere to reference it, but I don't know where. Does any of this make the issue clearer?

EDIT: By the way, if you can help me get this working I will love you for creating this awesome extension. Just sayin'. ^.^

**Comment by quetzalcoatl** [comment]
Indeed, those details are very important - especially the fact you are unable to run the 'installer utility'. This makes me think that I also messed up something in the references between modules.

---

EDIT: I've already checked some possibilities, and the problem with installer is not about modules, but is related to killing QTAgent that I've added "just in case" at some point of time. Please at first test the VSIX for which I've pasted a link far down below this lenghy post.

---

Your description of the project you are working on matches exactly what I thought, the Ctrl-R,A effects, too. I dont know about the installer, but the plugin is not running only because it "doesn't see" the assemblies that should have been copied at the first run by the 'installer utility'. If that was unable to run at all, then the "third bullet" is rubbish, as the correct xml-entries and files are nowhere to be found. Well, the files are there, but entries are not.

So, to repair the test running, we can currently only try to mimic the installer manually.

1) locate your 'visualstudio private assemblies directory'. This is where the VS keeps its some of its internal files, used very often but not meant to be just 'available to anyone'. It's base path is:

```
ProgramFiles\Microsoft Visual Studio 10.0\Common7\IDE\PrivateAssemblies
```

usually, it sits just right in the C:\Program Files. Let's call it (A)

2) locate your 'visualstudio extension directory'. This is where the VS keeps all the extension files that you have installed by VSIX packages. It's base path is:

```
AppData\Microsoft\VisualStudio\10.0\Extensions
```

but depending on your exact setup, it may land in a few different places. On Windows7, for a local user account, it should be:

```
C:\Users\ (YourUserAccountName) \AppData\Local\Microsoft\VisualStudio\10.0\Extensions
```

3) There, if the VSIX installed correctly, in there should be a further subfolder:

```
quetzalcoatl\xUnit Test Runner for VisualStudio 2010\2.1
```

let's call it (B). Inside it there you should find a bunch of DLL files. You will see the 'blahblah.autoinstaller.exe' - do not run it. At best, it will crash. It needs a lot of parameters to be run correctly. Normally, they are generated by the plugin when you launch it from "test" menu.

Ok, now to the point. The first way is hyper quick, but can break something in the future (drawbacks are listed at the end of the post):

4) from (B) copy the files right into (A):  
- xunit.runner.visualstudio.vs2010.core.dll  
- xunit.runner.utility.dll

And this is it. Everything should be now working, unless there are some more x64-specific yet unknown bugs. However, I really suggest you do it in the long way, because the installer will recognize it later once it is fixed.

The long way is:

4) in (A) create a subfolder "xunit.runner.visualstudio.vs2010"  
5) copy the above two files to this subfolder  
6) in (A) locate the following file(s):  
- QTAgent.exe.config  -- the file for testing 64bit code
- QTAgent_35.exe.config    -- the file for testing 64bit code under exactly .Net 3.5
- QTAgent32.exe.config    -- the file for testing 32bit code
- QTAgent32_35.exe.config    -- the file for testing 32bit code under exactly .Net 3.5

7) for either each of those files, or just for the file that corresponds to your setup:
7.1) edit that file in any relatively sane text editor; even notepad will be enough, but I wont guarantee for any ed-like commandline ancient things  
7.2) at the top of the file, you will see:  

```
<configuration>
  <runtime>
    <gcServer enabled="true" />
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <probing privatePath=" (...BLAHBLAHBLAH...) " />
    </assemblyBinding>
  </runtime>
```

immediately before the closing tag of "assemblyBinding", paste exactly that:

```
      <dependentAssembly>
        <assemblyIdentity name="xunit.runner.utility" publicKeyToken="8d05b1bb7a6fdb6c" culture="neutral" processorArchitecture="msil" />
        <codeBase version="2.0.0.0" href="PrivateAssemblies\xunit.runner.visualstudio.vs2010\xunit.runner.utility.dll" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="xunit.runner.visualstudio.vs2010.core" publicKeyToken="8caa1cd9f64d6a1a" culture="neutral" processorArchitecture="msil" />
        <codeBase version="2.1.0.0" href="PrivateAssemblies\xunit.runner.visualstudio.vs2010\xunit.runner.visualstudio.vs2010.core.dll" />
      </dependentAssembly>
```

so that your file's beginning should now look like:

```
<configuration>
  <runtime>
    <gcServer enabled="true" />
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <probing privatePath="(...BLAHBLAHBLAH...)" />
      <dependentAssembly>
        <assemblyIdentity name="xunit.runner.utility" (...BLAH...) />
        <codeBase version="2.0.0.0" href="(...BLAH...)\xunit.runner.utility.dll" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="xunit.runner.visualstudio.vs2010.core" (...BLAH...) />
        <codeBase version="2.1.0.0" href="(...BLAH...)\xunit.runner.visualstudio.vs2010.core.dll" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
```

and this is it. It is quite mindless and tedious job, so the file copying and config-editing is usually done automatically by the installer utility. It would apply exactly those changes, and also it would enable you to remove/uninstall these files and entries later when you get bored with xUnit.

If you follow the 'quick path', everything should be OK until you decide to remove or upgrade the xvsr10 extension. Placing the DLLs right into in the PrivateAssemblies makes them an **override** of those kept in the extension directory (B), and the VS may simply not notice that any files were upgraded. This means, that the quick way forces you to refresh the files manually at each upgrade (that is: copy again from (B) to (A)). It can also cause problems if you have many local users on the same machine and they want to use different versions of the plugin. Yeah, I know, that's just sooo very probable. :)

**Comment by quetzalcoatl** [comment]
Hope this helps. If the long way works for you, I'll try to patch the installer to cover all four .config files, not only the 'probable' one.

**Comment by quetzalcoatl** [comment]
Also, please do a small test for me: uninstall the extension by tools->extensionmanager, and try installing this one:

https://github.com/quetzalcoatl/xvsr10/blob/iss0001-x64-installation/dist/xunit.runner.visualstudio.vs2010.vsix?raw=true

In that file, I've added some error handling around the most probable failure point, and you should be able to run the installer. If you get some error messages, I'd be grateful for screenshots.

If the installer works now, then it will do the "copy files to subfolder" for you correctly. But, it will paste the required config sections to "QTAgent.exe.config" file instead of "QTAgent32.exe.config", and this you will still have to do manually. I think I'll manage to fix that tomorrow, but I need to know if the installer starts at all!

**Comment by Riobe** [comment]
Awesome! Thanks for the replies. I have no problem trying the long way, but I'll uninstall and try the new VSIX first. I'm at work at the moment so it'll be later, when I get home (I don't have any sort of VPN setup for my home machine).

**Comment by quetzalcoatl** [comment]
I've updated the https://github.com/quetzalcoatl/xvsr10/blob/iss0001-x64-installation/dist/xunit.runner.visualstudio.vs2010.vsix?raw=true yet again. It should now patch all .config files, too, so if it actually runs, no manual steps should be required.

Let me know if it worked for you!

**Comment by quetzalcoatl** [comment]
Have you tried that new package (that patching all cfg files) by chance?

**Comment by quetzalcoatl** [comment]
ping?

**Comment by Riobe** [comment]
Sorry, it's been awhile since I've had time to code or pay attention to coding emails. (been crazy for me lately) I tried with the new link and got the same result: The 'Xunit.Runner.VisualStudio.VS2010.XUnitTestPackage, xunit.runner.visualstudio.vs2010, Version=2.1.1.2, Culture=neutral, PublicKeyToken=8caa1cd9f64d6a1a' package did not load correctly.

I'm not sure what details might be useful to you at this point, but I'd be happy to look them up if you are still interested in telling me what they are.

**Comment by quetzalcoatl** [comment]
Damn :( In the next 2 weeks I should have some spare time, maybe I'll manage to get some x64 (maybe virtual?) machine to check that on my own. It seems I've guessed wrong, or overlooked some detail. Thanks again for rechecking this for me and for your help. I'll surely drop you a note once it is fixed for sure.

**Comment by Riobe** [comment]
If you want, we could set up a time to do a remote assistance and you could actually check it out on my PC if that would help. You might know what to look for better than I.

**Comment by jberezanski** [comment]
Similar problem here.
- Windows 7 64-bit
- Visual Studio 2010 Ultimate
- solution and all projects compiled to "Any CPU"
- in test settings (a .testsettings file), Hosts tab, bitness setting changed from "Force tests to run in 32 bit process" (the default) to "Run tests in 64 bit process on 64 bit machine"; these settings are selected as the active test settings

XVSR10 Module Manager displays "QTAgent32.exe.config" as the current QTAgent config, regardless of the active test settings. When running tests, however, Visual Studio starts the 64-bit QTAgent.exe, which is not configured with correct assembly bindings. 

I guess the mechanisms used by MSVST2A_Access.QTAgentExecutableFilename do not take the active test settings into account.

The vsix version posted above, which updates all configs, indeed resolves the problem. Since the agent bitness setting can be changed at will by the developer, I think patching all configs once and for all is the correct behavior for XVSR10 installer. It would be convenient if you released this as the official binary on the download page.

As a side note: great job! This extension actually motivated me to switch from NUnit to xUnit.

**Comment by quetzalcoatl** [comment]
Thanks for the note and reminder!  
Shamefully, I've almost completely forgot about the issue, since I'm still on 32bits and temporarily using eclipse :)

I'm glad the patch-all fix worked for you. The inner mechanisms indeed don't look at the run-config, because I think I had found it much harder to determine and keep in sync than to just patch all files. I'll check for forgotten commits and update binaries today in the evening.

**Comment by quetzalcoatl** [comment]
@Riobe: btw, is it still crashing for you or maybe you have found out what it was?

**Comment by quetzalcoatl** [comment]
well.. the 'evening' got stretched a bit, because I've had to cleanup a few things first. I've found out that one of the projects was build with x86 instead of AnyCPU, so maybe it was the problem for Riobe's problems. I've updated the https://github.com/downloads/quetzalcoatl/xvsr10/xunit.runner.visualstudio.vs2010.vsix to point to the newest 2.2 version that patches all cfg files. Also this version is now properly code-signed, both the binaries and the package it self :)
